private/pkg/manifest/manifest_test.go

```diff
@@ -0,0 +1,219 @@
+	retContent, err := m.MarshalText()
+	require.NoError(t, err)
+	assert.Equal(t, manifestContent, retContent, "round trip failed")
+}
+
+func TestEmptyManifest(t *testing.T) {
+	t.Parallel()
+	content, err := NewManifest().MarshalText()
+	assert.NoError(t, err)
+	assert.Equal(t, 0, len(content))
+}
+
+func TestAddContent(t *testing.T) {
+	t.Parallel()
+	// single entry
+	manifest := NewManifest()
+	err := manifest.AddContent("my/path", bytes.NewReader(nil))
+	require.NoError(t, err)
+	expect := fmt.Sprintf("%s  my/path\n", mkdigest(nil))
+	retContent, err := manifest.MarshalText()
+	require.NoError(t, err)
+	assert.Equal(t, expect, string(retContent))
+
+	// failing content read
+	expectedErr := errors.New("testing error")
+	err = manifest.AddContent("my/path", iotest.ErrReader(expectedErr))
+	assert.ErrorIs(t, err, expectedErr)
+}
+
+func testInvalidManifest(
+	t *testing.T,
+	desc string,
+	line string,
+) {
+	t.Helper()
+	t.Run(desc, func(t *testing.T) {
+		t.Parallel()
+		_, err := NewManifestFromReader(strings.NewReader(line))
+		assert.ErrorContains(t, err, desc)
+	})
+}
+
+func TestInvalidManifests(t *testing.T) {
+	testInvalidManifest(
+		t,
+		"invalid entry",
+		"\n",
+	)
+	testInvalidManifest(
+		t,
+		"invalid entry",
+		"whoops\n",
+	)
+	testInvalidManifest(
+		t,
+		"short digest",
+		"shake256:1234  foo\n",
+	)
+	testInvalidManifest(
+		t,
+		"unsupported hash",
+		"md5:d41d8cd98f00b204e9800998ecf8427e  foo\n",
+	)
+	testInvalidManifest(
+		t,
+		"malformed digest string",
+		"bar  foo\n",
+	)
+	testInvalidManifest(
+		t,
+		"encoding/hex",
+		"shake256:_  foo\n",
+	)
+	testInvalidManifest(
+		t,
+		"partial record",
+		fmt.Sprintf("%s  null", mkdigest(nil)),
+	)
+}
+
+func TestBrokenRead(t *testing.T) {
+	t.Parallel()
+	expected := errors.New("testing error")
+	_, err := NewManifestFromReader(iotest.ErrReader(expected))
+	assert.ErrorIs(t, err, expected)
+}
+
+func TestUnmarshalBrokenManifest(t *testing.T) {
+	t.Parallel()
+	var m Manifest
+	err := m.UnmarshalText([]byte("foo"))
+	assert.Error(t, err)
+}
+
+func TestFromBucket(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	bucket, err := storagemem.NewReadBucket(
+		map[string][]byte{
+			"null": nil,
+			"foo":  []byte("bar"),
+		})
+	require.NoError(t, err)
+	m, err := NewManifestFromBucket(ctx, bucket)
+	require.NoError(t, err)
+	expected := fmt.Sprintf("%s  foo\n", mkdigest([]byte("bar")))
+	expected += fmt.Sprintf("%s  null\n", mkdigest(nil))
+	retContent, err := m.MarshalText()
+	assert.NoError(t, err)
+	assert.Equal(t, expected, string(retContent))
+}
+
+func TestPaths(t *testing.T) {
+	t.Parallel()
+	m := NewManifest()
+	err := m.AddContent("path/one", bytes.NewReader(nil))
+	require.NoError(t, err)
+	err = m.AddContent("path/two", bytes.NewReader(nil))
+	require.NoError(t, err)
+	paths, ok := m.Paths(mkdigest(nil))
+	assert.True(t, ok)
+	assert.ElementsMatch(t, []string{"path/one", "path/two"}, paths)
```

Comment:
Nit: Add an assertion making sure they're sorted? or does the `ElementsMath` checks order too? If so, may be add them _unsorted_?

Reply:
`ElementsMatch` is an unordered, set equality assertion: 

> ElementsMatch asserts that the specified listA(array, slice...) is equal to specified listB(array, slice...) ignoring the order of the elements. If there are duplicate elements, the number of appearances of each of them in both lists should match.

Order is not guaranteed by `Paths` since it was never said to be, but I'll make that explicit in its docs.